clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -1035,6 +1043,13 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
     }
   }
 
+  // Re-link against any bitcodes supplied via the -mlink-builtin-bitcode option
+  // Some optimizations may generate new function calls that would not have
+  // been linked pre-optimization (i.e. fused sincos calls generated by
+  // AMDGPULibCalls::fold_sincos.)
+  if (ClRelinkBuiltinBitcodePostop)
```

Comment:
So, what I had in mind is that we could make a new `clang` option similar to `-mlink-builtin-bitcode`. This would then be used by the HIP toolchain or similar when constructing the list of files to pass via `-mlink-builtin-bitcode`. We would then simply register those with this secondary pass.

This approach seems much simpler, being a boolean option that just relinks everything, but I somewhat like the idea of `-mlink-builtin-bitcode` being a pre-link operation and having another one for post-linking. That being said, it may not be worth the extra work because this is a huge hack around this ecosystem already.

Reply:
I did look into having a file-specific option to link specific bitcodes post-optimization, and I think it is a good idea overall.

It is possible to come up with a specific set of bitcodes that need to be re-linked, based on the current optimizations we've implemented that introduce undefined functions. And we could then specify that set for a file-specific post-optimization linking.

But there are some downsides:
  - Any users would also need to know the specific set of bitcodes needed for post-optimization linking
  - We may need to hardcode that list in a bunch of different places, both within LLVM and external APIs
  - If a new optimization is implemented that expands the set of bitcodes needed for post-optimization linking, we'd need to go around and update any place that uses the file-specific options

Given that, I think the boolean relink option may serve us better for the time being

Reply:
Also, the reason for LinkInModules pass is that we need to insert the second linking step into the RunOptimizationPipeline, specifically _after_ the AMDGPULibCalls pass, but _before_ the BitcodeWriter pass

Reply:
It's definitely the easier option. This problem is pretty specific, but I can see it being easier to just remove this class of bugs entirely. It's an ugly solution for an ugly problem overall.